macOS/debugger.cpp

```diff
@@ -68,12 +68,67 @@ vm_prot_t Debugger::MacOSProtectionFlags(MemoryProtection memory_protection) {
   }
 }
 
+void Debugger::ClearSharedMemory() {
+  for (int i=0; i<shared_memory.size(); ) {
+    SharedMemory sm = shared_memory.at(i);
+    FreeSharedMemory(sm, i);
+  }
+
+  shared_memory.clear();
+}
+
+void Debugger::FreeSharedMemory(SharedMemory sm, int index) {
+  if (sm.size == 0) {
+    WARN("FreeShare is called with size == 0\n");
+    return;
+  }
+
+  kern_return_t krt = mach_port_destroy(mach_task_self(), sm.port);
+  if (krt != KERN_SUCCESS) {
+    FATAL("Error (%s) destroy port for local shared memory @ 0x%llx\n", mach_error_string(krt), sm.local_address);
+  }
+
+  krt = mach_vm_deallocate(mach_task_self(), sm.remote_address, sm.size);
+  if (krt != KERN_SUCCESS) {
+    FATAL("Error (%s) freeing memory @ 0x%llx\n", mach_error_string(krt), sm.remote_address);
+  }
+
+  std::list<SharedMemory>::iterator it;
+  if (index == -1) {
+    shared_memory.erase(std::remove(shared_memory.begin(), shared_memory.end(), sm), shared_memory.end());
+  } else {
+    shared_memory.erase(shared_memory.begin() + index);
```

Comment:
Since we are potentially removing elements from the middle, a vector might not be the most suitable datastructure here. I think a list would work better, and I suggest renaming

```
void Debugger::FreeSharedMemory(SharedMemory sm, int index)
```

into 

```
void Debugger::FreeSharedMemory(list<SharedMemory>::iterator &it)
```

which I also think is cleaner than passing both SharedMemory and an index.

Reply:
Ah yeah, I also think about this. I tried to pass iterator, but what if the use case is that user want to free shared memory without iterating through `shared_memory` list?

Reply:
Yep, for now that's OK. Alternately, you could break it into two functions like this:
```
void Debugger::FreeSharedMemory(list<SharedMemory>::iterator &it);
void Debugger::FreeSharedMemory(SharedMemory &sm);
```
Where the first one would call the second one and also remove the entry from the list.